### PR TITLE
Terrain Supports Physics test

### PR DESCRIPTION
**Pull request by jjjoness**
Signed-off-by: John Jones-Steele <82226755+jjjoness@users.noreply.github.com>

### CODE REVIEW

**Review by moraaar — APPROVED** [review]
LGTM

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        checkID = pte.get_container_item("Configuration|Gradient Entities", 0)
+        Report.result(Tests.entity_added, checkID.GetValue() == test_entity2.id)
+
+        # 8) Set the PhysX Collider to Sphere mode
+        shape = 0
+        hydra.get_set_test(ball, 1, "Shape Configuration|Shape", shape)
+        setShape = hydra.get_component_property_value(ball.components[1], "Shape Configuration|Shape")
+        Report.result(Tests.shape_set, shape == setShape)
+
+        # 9) DIsable and Enable the Terrain Gradient List so that it is recognised
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:120]: Yes, it first disables all components before re-enabling them. Comment changed.

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    Summary:
+    Test aspects of the TerrainHeightGradientList through the BehaviorContext and the Property Tree.
+
+    Test Steps:
+    Expected Behavior:
+    The Editor is stable there are no warnings or errors.
+
+    Test Steps:
+     1) Load the base level
+     2) Create 2 test entities, one parent at 521.0, 512.0, 50.0 and one child at the default position
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:33]: Fixed

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    
+    # 1) Load the level
+    helper.open_level("", "Base")
+    helper.wait_for_condition(lambda: general.get_current_level_name() == "Base", 2.0)
+
+    #1a) Load the level components
+    hydra.add_level_component("Terrain World")
+    hydra.add_level_component("Terrain World Renderer")
+
+    # 2) Create 2 test entities, one parent at 521.0, 512.0, 50.0 and one child at the default position
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:73]: Fixed

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    entity1_components_to_add = ["Axis Aligned Box Shape", "Terrain Layer Spawner", "Terrain Height Gradient List", "Terrain Physics Heightfield Collider", "PhysX Heightfield Collider"]
+    entity2_components_to_add = ["Vegetation Reference Shape", "Gradient Transform Modifier", "FastNoise Gradient"]
+    ball_components_to_add = ["Sphere Shape", "PhysX Collider", "PhysX Rigid Body"]
+    test_entity1 = hydra.Entity("TestEntity1")
+    test_entity1.create_entity(azmath.Vector3(512.0, 512.0, 46.0), entity1_components_to_add)
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:78]: Commented in the code.

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    Report.result(Tests.create_test_entity1, test_entity1.id.IsValid())
+    test_entity2 = hydra.Entity("TestEntity2")
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:80]: Done.

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+    #1a) Load the level components
+    hydra.add_level_component("Terrain World")
+    hydra.add_level_component("Terrain World Renderer")
+
+    # 2) Create 2 test entities, one parent at 521.0, 512.0, 50.0 and one child at the default position
+    entity1_components_to_add = ["Axis Aligned Box Shape", "Terrain Layer Spawner", "Terrain Height Gradient List", "Terrain Physics Heightfield Collider", "PhysX Heightfield Collider"]
+    entity2_components_to_add = ["Vegetation Reference Shape", "Gradient Transform Modifier", "FastNoise Gradient"]
+    ball_components_to_add = ["Sphere Shape", "PhysX Collider", "PhysX Rigid Body"]
+    test_entity1 = hydra.Entity("TestEntity1")
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:77]: Done.

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    Test aspects of the TerrainHeightGradientList through the BehaviorContext and the Property Tree.
+
+    Test Steps:
+    Expected Behavior:
+    The Editor is stable there are no warnings or errors.
+
+    Test Steps:
+     1) Load the base level
+     2) Create 2 test entities, one parent at 512.0, 512.0, 50.0 and one child at the default position and add the required components
+     2a) Create a ball at 600.0, 600.0, 46.0 - This position is bot too high over the heighfield so will collide in a reasonable time
```
Comment (greerdv) [inline AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:34]: nitpick - looks like a typo in "bot"

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+     3) Start the Tracer to catch any errors and warnings
+     4) Change the Axis Aligned Box Shape dimensions
+     5) Set the Vegetation Shape reference to TestEntity1
+     6) Set the FastNoise gradient frequency to 0.01
+     7) Set the Gradient List to TestEntity2
+     8) Set the PhysX Collider to Sphere mode
+     9) Disable and Enable the Terrain Gradient List so that it is recognised
```
Comment (greerdv) [inline AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:41]: Is this something a user has to do in the editor or is it only required in the test environment for some reason? If you do have to do it in the editor, doesn't seem like an ideal workflow.

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    entity1_components_to_add = ["Axis Aligned Box Shape", "Terrain Layer Spawner", "Terrain Height Gradient List", "Terrain Physics Heightfield Collider", "PhysX Heightfield Collider"]
+    entity2_components_to_add = ["Vegetation Reference Shape", "Gradient Transform Modifier", "FastNoise Gradient"]
+    ball_components_to_add = ["Sphere Shape", "PhysX Collider", "PhysX Rigid Body"]
+    terrain_spawner_entity = hydra.Entity("TestEntity1")
+    terrain_spawner_entity.create_entity(azmath.Vector3(512.0, 512.0, 50.0), entity1_components_to_add)
+    Report.result(Tests.create_terrain_spawner_entity, terrain_spawner_entity.id.IsValid())
+    height_provider_entity = hydra.Entity("TestEntity2")
+    height_provider_entity.create_entity(azmath.Vector3(0.0, 0.0, 0.0), entity2_components_to_add,terrain_spawner_entity.id)
+    Report.result(Tests.create_height_provider_entity, height_provider_entity.id.IsValid())
+    # 2a) Create a ball at 600.0, 600.0, 46.0 - This position is bot too high over the heighfield so will collide in a reasonable time
```
Comment (greerdv) [inline AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:83]: nitpick - same typo again

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+        Report.result(Tests.shape_changed, entityId == terrain_spawner_entity.id)
+
+        # 6) Set the FastNoise Gradient frequency to 0.01
+        Frequency = 0.01
+        height_provider_entity.get_set_test(2, "Configuration|Frequency", Frequency)
+        FrequencyVal = hydra.get_component_property_value(height_provider_entity.components[2], "Configuration|Frequency")
+        Report.result(Tests.frequency_changed, math.isclose(Frequency, FrequencyVal, abs_tol = 0.00001))
+
+        # 7) Set the Gradient List to TestEntity2
+        pte = hydra.get_property_tree(terrain_spawner_entity.components[2])
```
Comment (greerdv) [inline AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:110]: would it be possible to rename this variable to something more explanatory?

File: AutomatedTesting/Levels/Base/Base.prefab [hunk 11]
```diff
@@ -17,7 +17,16 @@
             },
             "Component_[14126657869720434043]": {
                 "$type": "EditorEntitySortComponent",
-                "Id": 14126657869720434043
```
Comment (greerdv) [inline AutomatedTesting/Levels/Base/Base.prefab:20]: Is it intended to check this change in? Can it affect any of the other levels which load the base prefab?

**Review by greerdv — APPROVED** [review]

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py (hunk 7 above)
Comment (jckand-amzn) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:34]: And in "heightfield".

File: AutomatedTesting/Gem/PythonTests/Terrain/TestSuite_Main.py [hunk 12]
```diff
@@ -19,7 +19,9 @@
 @pytest.mark.parametrize("launcher_platform", ['windows_editor'])
 @pytest.mark.parametrize("project", ["AutomatedTesting"])
 class TestAutomation(EditorTestSuite):
-    #global_extra_cmdline_args=["--regset=/Amazon/Preferences/EnablePrefabSystem=true"]
 
     class test_AxisAlignedBoxShape_ConfigurationWorks(EditorSingleTest):
         from .EditorScripts import TerrainPhysicsCollider_ChangesSizeWithAxisAlignedBoxShapeChanges as test_module
+
+    class test_Terrain_SupportsPhysics(EditorSingleTest):
```
Comment (jckand-amzn) [inline AutomatedTesting/Gem/PythonTests/Terrain/TestSuite_Main.py:26]: Have you attempted to run these 2 tests together as EditorSharedTest so they're batched? I don't see any reason it wouldn't work, but maybe missed it.

**Review by jckand-amzn — COMMENTED** [review]
I know this is already merged, but would be worth taking a look at the batching as it would cut the runtime of the terrain test suite by ~50%.

File: AutomatedTesting/Levels/Base/Base.prefab (hunk 11 above)
Comment (jjjoness) [reply AutomatedTesting/Levels/Base/Base.prefab:20]: There wasn't a Base.prefab, I was told to create one.

File: AutomatedTesting/Gem/PythonTests/Terrain/TestSuite_Main.py (hunk 12 above)
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/TestSuite_Main.py:26]: I'll test this and change if it works.

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py (hunk 8 above)
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:41]: Will fix.

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py (hunk 9 above)
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:83]: Will fix.

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py (hunk 10 above)
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:110]: Will fix.

File: AutomatedTesting/Gem/PythonTests/Terrain/TestSuite_Main.py (hunk 12 above)
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/TestSuite_Main.py:26]: Just tested and they work fine. I'll setup a quick PR in the morning to add these changes.

Comment (jckand-amzn) [reply AutomatedTesting/Gem/PythonTests/Terrain/TestSuite_Main.py:26]: Awesome!

File: AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py (hunk 7 above)
Comment (jjjoness) [reply AutomatedTesting/Gem/PythonTests/Terrain/EditorScripts/Terrain_SupportsPhysics.py:34]: Fixed.

**Pull request merged**